feat: webhooks docs (#531)
momentohq/public-dev-docs: docs/topics/develop/api-reference/webhooks.md

---
sidebar_position: 6
title: Webhook API reference information
sidebar_label: Webhooks
description: Learn the Webhook API calls you need to know about and how to use them with Momento services.
---

# Webhook API reference

<img src="/img/topics/webhooks-page.png" width="90%" alt="minified javascript code with the text 'Trigger the hook'" />

> Webhooks provide an additional way to consume messages on a topic. When a webhook is created, it gets attached to a topic. When messages are published to this topic, an outbound http call gets POSTed to the webhook callback url. The endpoint that is associated with the webhook must be publicly accessible for Momento to be able to POST requests. To confirm receipt, your api should respond with an empty HTTP 200 response to the original request. If confirmation is not received within _5 seconds_, we drop the message and mark it as timed out.  

Reviewer: Any HTTP 2xx is fine, it doesn't have to be a 200 exactly.

> Webhooks provide an additional way to consume messages on a topic. When a webhook is created, it gets attached to a topic. When messages are published to this topic, an outbound http call gets POSTed to the webhook callback url. The endpoint that is associated with the webhook must be publicly accessible for Momento to be able to POST requests. To confirm receipt, your api should respond with an empty HTTP 200 response to the original request. If confirmation is not received within _5 seconds_, we drop the message and mark it as timed out.  

Author: ```suggestion
Webhooks provide an additional way to consume messages on a topic. When a webhook is created, it gets attached to a topic. When messages are published to this topic, an outbound http call gets POSTed to the webhook callback url. The endpoint that is associated with the webhook must be publicly accessible for Momento to be able to POST requests. To confirm receipt, your api should respond with an empty HTTP 2xx response to the original request. If confirmation is not received within _5 seconds_, we drop the message and mark it as timed out.  
```

| postUrl  | string | The URL to which the payload will be sent using HTTP POST. |

## FAQ

<details>
<summary>How do I validate incoming webhook requests?</summary>

Each webhook that gets created will get a unique signing secret. Using this signing secret, you can verify whether this request came from Momento. On each HTTP request that Momento sends, we include a `momento-signature` header. This signature is created by combining the the signing secret with the request body using a standard HMAC hash.

Steps to validate

1. Retrieve the `momento-signature` header from the request, as well as the signing secret for the webhook (stored in a secret vault or environment variable)
2. Using HMAC SHA3-256, hash the request body, using the signing secret associated with the webhook
3. Compare the computed signature to the `momento-signature` header on the request

```typescript
import * as crypto from 'crypto';

const didRequestComeFromMomento = (req: Request): boolean => {
  const hash = crypto.createHmac("SHA3-256", "the signing secret");
  const hashed = hash.update(req.rawBody).digest('hex');
  return hashed === req.headers['momento-signature'];
}

> ```

Reviewer: oh neat. I"m glad we have both server + client code for this.

</details>

<details>
<summary>What is the structure of the POST body?</summary>

```typescript
{
  cache: string;
  topic: string;
  event_timestamp: number;
  publish_timestamp: number;
  topic_sequence_number: number;
  token_id: string; // is an empty string if the message was published with an api key that does not contain a token_id
  text: string; // this is the message that was published
}
```

</details>

<details>
<summary>What should I do if the webhook subscription fails?</summary>

> Check the provided error response for details. Common issues include invalid cache or topic names, or issues with the provided `postUrl`.

Reviewer: I don't think there is an error response though for failed Webhook calls. There'll be metrics shortly after launch.

> Check the provided error response for details. Common issues include invalid cache or topic names, or issues with the provided `postUrl`.

Author: ```suggestion
Ensure that your webhook url is public and valid. Common issues include wrong cache or topic names, or issues with the provided `postUrl`.
```

| webhookName | string | The name of the webhook.                    |

## WebhookDestination object

The `WebhookDestination` object contains the information about where the webhook payload will be sent.

| Name     | Type   | Description                       |
|----------|--------|-----------------------------------|
| postUrl  | string | The URL to which the payload will be sent using HTTP POST. |

## FAQ

<details>
<summary>How do I validate incoming webhook requests?</summary>

Each webhook that gets created will get a unique signing secret. Using this signing secret, you can verify whether this request came from Momento. On each HTTP request that Momento sends, we include a `momento-signature` header. This signature is created by combining the the signing secret with the request body using a standard HMAC hash.

Steps to validate

1. Retrieve the `momento-signature` header from the request, as well as the signing secret for the webhook (stored in a secret vault or environment variable)
2. Using HMAC SHA3-256, hash the request body, using the signing secret associated with the webhook
3. Compare the computed signature to the `momento-signature` header on the request

> ### Example

Author: ```suggestion
```
